### PR TITLE
[REVIEW] Port nvstrings Regular Expressions functions to cuDF Python/Cython

### PR DESCRIPTION
**Pull request by galipremsagar**
Resolves: #3989

Ported:

- [x] .str.contains
- [x] .str.count
- [x] .str.extract
- [x] .str.replace (.str.replace & .str.replace_multi as .str.replace)
- [x] .str.findall
- [x] .str.replace_with_backrefs
- [x] Re-visit existing tests/Write new tests.
- [x] Update Docs.

### CODE REVIEW

File: python/cudf/cudf/_libxx/strings/replace_re.pyx [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    regular expressions `patterns` with their corresponding
+    strings in `repl_strings` in `source_strings`.
+    """
+    cdef unique_ptr[column] c_result
+    cdef column_view source_view = source_strings.view()
+    cdef column_view repl_view = repl_strings.view()
+
+    cdef vector[string] patterns_vector
+    for pattern in patterns:
+        patterns_vector.push_back(str.encode(pattern))
```
[the remark this answers is not in the document]
Comment (galipremsagar) [reply python/cudf/cudf/_libxx/strings/replace_re.pyx]: I tried allocating size while initialization like this:
```
cdef vector[string] patterns_vector
patterns_vector = vector[string](2)
```
and also like this:
```
cdef vector[string] *patterns_vector= new vector[string](2)
```
It continues to compile but hangs up during run-time. Tried this too: https://stackoverflow.com/a/34833034
The stack-overflow solution doesn't seem to work incase of `string`, Browsed [through docs](https://cython.readthedocs.io/en/latest/src/userguide/wrapping_CPlusPlus.html) of cython but didn't find a way to reserve for strings.

[the remark this answers is not in the document]
Comment (kkraus14) [reply python/cudf/cudf/_libxx/strings/replace_re.pyx]: Those are both constructors, you just want to reserve: https://en.cppreference.com/w/cpp/container/vector/reserve
```
cdef vector[string] patterns_vector
patterns_vector.reserve(2)
```

[the remark this answers is not in the document]
Comment (galipremsagar) [reply python/cudf/cudf/_libxx/strings/replace_re.pyx]: Thanks @kkraus14, That worked.

**Comment by jlowe** [comment]
Noticed this has been marked Ready to Merge, but it looks like there was an upmerge issue with branch-0.13.  A number of commits would be reverted if this was merged.

**Comment by galipremsagar** [comment]
> Noticed this has been marked Ready to Merge, but it looks like there was an upmerge issue with branch-0.13. A number of commits would be reverted if this was merged.

Yes, resolving it.

**Comment by kkraus14** [comment]
My github UI did not show the changes of course when I approved 🤦‍♂

**Pull request merged**